### PR TITLE
[AutoDiff] Add more `Tensor` op adjoint definitions.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add adjoint definitions for:
- `Tensor.squared`
- `Tensor.transposed` (all variants)
- `sigmoid`
- `softmax`
- `logSoftmax`

Redundant adjoint definitions for op variants (e.g. all the variants of
`transposed`) can be removed once AD supports generics.

### CODE REVIEW

**Comment by dan-zheng** [comment]
Adding some runtime tests.

File: stdlib/public/TensorFlow/Gradients.swift [hunk 1]
```diff
@@ -355,10 +355,19 @@ func _adjointRsqrt<T : Differentiable & FloatingPoint>(
   return -seed / 2 * pow(originalValue, 3)
 }
 
-func _adjointSquared<T : Differentiable & FloatingPoint>(
+@inlinable
+func _adjointLogSoftmax<T : Differentiable & FloatingPoint>(
   _ seed: Tensor<T>, _ originalValue: Tensor<T>, _ x: Tensor<T>
 ) -> Tensor<T> {
-  return 2 * x * seed
+  let seed = seed.broadcast(like: originalValue)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Gradients.swift]: We should no longer broadcast seed to the original value in any derivative.

**Review by rxwei — APPROVED** [review]

File: stdlib/public/TensorFlow/Gradients.swift (hunk 1 above)
Comment (dan-zheng) [reply stdlib/public/TensorFlow/Gradients.swift]: I see - instead, we should assume that seed has same shape as original result and require users to specify non-scalar initial seeds via pullbacks?

That seems mathematically correct, tests using `gradient` that assume the scalar seed `Tensor(1)` gets broadcasted need to be rewritten using `pullback`.

Comment (rxwei) [reply stdlib/public/TensorFlow/Gradients.swift]: Correct!

**Comment by rxwei** [comment]
CI is down. Manually tested and all tests are passing. Merging to unblock progress.
